Simplify long if-chains using dispatch tables

php1ic opened the pull request:
<!-- Thank you for contributing a Pull Request! Please ensure you also check the contributor guidelines:
https://github.com/radioactivedecay/radioactivedecay/blob/main/CONTRIBUTING.md -->

#### What does this PR implement/fix?
There are long if-chains in two places in the Inventory class. This PR attempts to simplify and remove repetition whilst also making future additions easier to implement via [dispatch tables](https://martinheinz.dev/blog/90).

#### Fixes Issue
N/A

#### Other comments?
I think this PR makes the code slightly more difficult to reason out on first read, but believe it makes it more maintainable.

php1ic committed:
Use a dispatch table to simplify the long if-chain

This removes the repetition where only fairly minor things change and
should allow simple addition of future functionality.

php1ic committed:
Create another dispatch table to simplify if-chain

This one helps with the plotting and is slightly more complicated than
the one used in the previous commit (c96f3a42), but uses the same
principle to remove repetition.

php1ic committed:
Update formatting

No actual code change, a new black rule must have been added so allow it
to modify the file.

php1ic committed:
Check for key rather than catching an exception

Later, in a separate function, we check a key exists and raise if it
doesn't rather than catching an exception. Let's do the same here.

alexmalins commented:
Hi @php1ic - thanks for putting your thoughts into this :raised_hands: And sorry it took me a couple of weeks to take a look :bow: 

I'm happy to approve this :heavy_check_mark: Dictionary dispatching is definitely a better pattern than the long if else chains :see_no_evil: 

My q's are, given you acknowledge the merit of improved maintainability at a cost of some readability, was there a reason you chose three different ways of setting up the dispatch dictionaries:
1. tuple of method name (str) and bool if the method needs a units string passed
2. functions which return lambda functions
3. lambda functions

Would it be easier to read if you were consistent in using one of those three options throughout? Ex.

```
decay_unit_dispatch = (
    {
        unit: (
            lambda inv: inv.activities(unit),
            f"Activity ({unit})",
        )
        for unit in self._get_unit_converter().activity_units,
    }
    | {
        "num": (lambda inv: inv.numbers(), "Number of atoms"),
    },
)
```

I think maybe being consistent everywhere and using more explicit var names (e.g. u -> unit, d -> inv in lambdas) would help readability. Certainly I found implementation 1. for `decay_unit_dispatch` code change was the hardest to read initially, due to the boolean and the use of `getattr` which I had to go away and lookup :sweat_smile:

php1ic committed:
Refactor dispatch dictionary into class member

By consistently using lambdas, paired with label strings, we can
create a common dictionary and convert it to a member variable
allowing simple reuse. This also allows us to validate the requested
decay string and exit immediately, before any unnecessary work is
carried out.

php1ic commented:
No problem, thanks for the comments.

Reading the code again, I did get slightly obsessed with minimising the amount of code so agree with your suggestion of consistency. I've moved the dispatch dictionary to be a member of the class and it now has a tuple of a lambda and a string as the value for all units, so can be used in both of the previous locations.

I got stuck on failing tests due to [late binding closures](https://docs.python-guide.org/writing/gotchas/#late-binding-closures) which is why those lambdas that take arguments now have the additional parameter with a default value.

The relocation of dictionary to a member variable also allows us to check the requested unit at the start of the function and error out early without doing unnecessary work.
